src/Symfony/Component/JsonStreamer/Read/PhpAstBuilder.php

```diff
@@ -382,7 +382,7 @@ private function buildCollectionNodeStatements(CollectionNode $node, bool $decod
         $prepareDataStmts = $decodeFromStream ? [
             new Expression(new Assign($this->builder->var('data'), $this->builder->staticCall(
                 new FullyQualified(Splitter::class),
-                $node->getType()->isList() ? 'splitList' : 'splitDict',
+                !$node->getType()->getCollectionKeyType()->isIdentifiedBy(TypeIdentifier::STRING) ? 'splitList' : 'splitDict',
```

Comment:
why not this? what about `iterable<Foo, Bar>` (eg from SplObjectStorage)?

```suggestion
                $node->getType()->getCollectionKeyType()->isIdentifiedBy(TypeIdentifier::INT) ? 'splitList' : 'splitDict',
```

Reply:
By doing this, `iterable<int|string, Foo>` will be encoded with `splitList`, which is not the intended behavior.
But I didn't think about `SplObjectStorage`, thanks! Updated consequently.